nestkernel/node_collection.cpp

```diff
@@ -165,7 +165,7 @@ NodeCollection::create_( const std::vector< index >& node_ids )
 
   std::vector< NodeCollectionPrimitive > parts;
 
-  index old_node_id = 0;
+  index old_node_id = current_first;
   for ( auto node_id = ++( node_ids.begin() ); node_id != node_ids.end(); ++node_id )
   {
     if ( *node_id == old_node_id )
```

Comment:
This test compares only neighboring entries in `node_ids` and does not enforce that they are strictly increasing. Therefore, ill-formed inputs can lead to unexpected results, e.g.
```Python
In [6]: nest.NodeCollection([1, 2, 3, 1, 5])                                                                                                                                              
Out[6]: 
NodeCollection(metadata=None,
               model=iaf_psc_alpha, size=1, first=1;
               model=iaf_psc_alpha, size=3, first=1, last=3;
               model=iaf_psc_alpha, size=1, first=5)
```
To catch both problems, I'd suggest `*node_id <= old_node_id`.
Btw, I am wondering if `auto& node_id` would save you dereferencing `*`s inside the loop.

Reply:
@heplesser I can't seem to reproduce your result, for me it throws an error. The `node_ids` argument in `create_` is expected to be sorted, so it shouldn't be a problem. However the sortedness is never checked, but I can add an `assert` for that.

And `node_id` is an iterator which needs to be dereferenced. But I can rename it to for example `node_id_it` to make it more clear.

Reply:
@hakonsbm My mistake! My example above was based on master, not your branch. Also ignore my comment on the iterator!

But I have another issue: 
- If `create_()` expects its input to be sorted, this should be clearly documented somewhere.
- The other questions is if the input that `create_()` receives should at all be sorted on its way from the Python level down to `create_()`, see also the discussions in #1563 and #1570. I would tend to leave the sorting to the user; that saves the sorting in the NEST code, and we would here just have to have a little and very lightweight check for strict monotony.

Reply:
@heplesser I can add documentation on the `create_()` function, and I agree that we can leave the sorting to the user. But is that within the scope of this PR, or would it be better to create a separate one?

Reply:
@hakonsbm Let us set this one on hold until we have come to a conclusion in #1570.

Reply:
In #1570 we decided that the nodes should be ordered, documentation has been added, so I am resolving this conversation.